packages/flutter_tools/lib/src/base/context.dart

```diff
@@ -3,79 +3,186 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:collection';
+
+import 'package:meta/meta.dart';
+
+/// Generates an [AppContext] value.
+///
+/// Generators are allowed to return `null`, in which case the context will
+/// store the `null` value as the value for that type.
+typedef dynamic Generator();
+
+/// An exception thrown by [AppContext] when you try to get a [Type] value from
+/// the context, and the instantiation of the value results in a dependency
+/// cycle.
+class ContextDependencyCycleException implements Exception {
+  ContextDependencyCycleException._(this.cycle);
+
+  /// The dependency cycle (last item depends on first item).
+  final List<Type> cycle;
+
+  @override
+  String toString() {
+    final StringBuffer buf = new StringBuffer('Dependency cycle detected: ');
+    buf.write(cycle.join(' -> '));
+    return buf.toString();
```

Comment:
nit: this is the same as: `=> 'Dependency cycle detected: ${cycle.join(' -> ')}'`